src/main/java/com/puppycrawl/tools/checkstyle/checks/design/FinalClassCheck.java

```diff
@@ -40,9 +40,34 @@
 
 /**
  * <p>
- * Checks that a class that has only private constructors and has no descendant
- * classes is declared as final.
+ * Check that identifies classes that can be declared as final are explicitly made final.
+ * The following is the type of classes that can be identified:
  * </p>
+ * <ol>
+ *   <li>
+ *       Private classes with implicit constructor.
+ *   </li>
+ *   <li>
+ *       Classes with all the type of modifier which contains only private constructor.
+ *   </li>
+ * </ol>
+ * <p>
+ * Classes are skipped if:
+ * </p>
+ * <ol>
+ *   <li>
+ *       Class is already Final.
+ *   </li>
+ *   <li>
+ *       Class is declared as abstract.
+ *   </li>
+ *   <li>
+ *       Class is Super class of some Anonymous inner class.
+ *   </li>
+ *   <li>
+ *       Class is with nested sub class.
```

Comment:
Just a question and not asking for a change, why are we skipping over classes with sub-classes? I know this was how it was before, but is there a reason this is?

Reply:
```
    private class SuperClass { // ok
        private SuperClass() {
        }
    }

    private class SubClass extends SuperClass {
    } // violation above 'Class SubClass should be declared as final'

```
I hope i have not make mistake in wording i want to show the above example. if `SuperClass` is final then we can't inherit it

Reply:
But this says nested subclass. Wouldn't that be something like: (example may not be made compilable) 
````
    private class SuperClass { // ok
        private SuperClass() {
        }

      private class SubClass extends SuperClass {
      }
    }


````

Reply:
ok I think you are telling about current situation 
```
kevin@inspiron-15-5510:~/Desktop/check_style$ cat config.xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC "-//Puppy Crawl//DTD Check Configuration 1.3//EN" "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">

<module name="Checker">
    <module name="TreeWalker">
        <module name="FinalClass"></module>
    </module>
</module>
kevin@inspiron-15-5510:~/Desktop/check_style$ cat Test.java
public class Test {

    private  class SuperClass { // ok
        private SuperClass() {
        }

      private class SubClass extends ss {
      }
    }

}

class ss {}
kevin@inspiron-15-5510:~/Desktop/check_style$ java -jar /home/kevin/Downloads/checkstyle-10.11.0-all.jar -c config.xml Test.java
Starting audit...
[ERROR] /home/kevin/Desktop/check_style/Test.java:3:5: Class SuperClass should be declared as final. [FinalClass]
Audit done.
Checkstyle ends with 1 errors.

------------------------------------------

kevin@inspiron-15-5510:~/Desktop/check_style$ java -jar /home/kevin/Desktop/check_style/checkstyle/checkstyle/target/checkstyle-10.12.0-SNAPSHOT-all.jar -c config.xml Test.java
Starting audit...
[ERROR] /home/kevin/Desktop/check_style/Test.java:3:5: Class SuperClass should be declared as final. [FinalClass]
[ERROR] /home/kevin/Desktop/check_style/Test.java:7:7: Class SubClass should be declared as final. [FinalClass]
Audit done.
Checkstyle ends with 2 errors.
kevin@inspiron-15-5510:~/Desktop/check_style$ 
```

Reply:
But isn't your example being triggered because of the other rules? `SuperClass` should be `final` since it has a private constructor. `SubClass` because it is a `private` class with the default constructor.

Reply:
You are right the example I have given is triggered due to other rules. 

@rnveach The thing I want to mention with `Class is with nested subclass.` is that we cannot show violation if a superclass is inherited in a subclass that is nested otherwise it becomes non-compilable.
```
kevin@inspiron-15-5510:~/Desktop/check_style$ cat config.xml 
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC "-//Puppy Crawl//DTD Check Configuration 1.3//EN" "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">

<module name="Checker">
    <module name="TreeWalker">
        <module name="FinalClass"></module>
    </module>
</module>

kevin@inspiron-15-5510:~/Desktop/check_style$ cat Test.java
public class Test {
    private class SuperClass { // ok 
        private SuperClass() {
        }

        private class SubClass extends SuperClass { // violation
        }
    }
}
kevin@inspiron-15-5510:~/Desktop/check_style$ 
kevin@inspiron-15-5510:~/Desktop/check_style$ java -jar /home/kevin/Downloads/checkstyle-10.11.0-all.jar -c config.xml Test.java
Starting audit...
Audit done.

-----------

kevin@inspiron-15-5510:~/Desktop/check_style$ java -jar /home/kevin/Desktop/check_style/checkstyle/checkstyle/target/checkstyle-10.12.0-SNAPSHOT-all.jar -c config.xml Test.java
Starting audit...
[ERROR] /home/kevin/Desktop/check_style/Test.java:6:9: Class SubClass should be declared as final. [FinalClass]
Audit done.
Checkstyle ends with 1 errors.
kevin@inspiron-15-5510:~/Desktop/check_style$ 
```

Reply:
> we cannot show violation if a superclass is inherited in a subclass

But the documentation you wrote does not say this. It just says a class with nested sub-classes.

Is it actually coded to look for "a superclass is inherited in a subclass", or are you saying that is what it was intended to cover?

Reply:
My purpose was to show this only `superclass is inherited in a subclass`.
for good wording, I have directly https://github.com/checkstyle/checkstyle/blob/6f451fbb49afc21b0e04b46c0576c7a984419a88/src/main/java/com/puppycrawl/tools/checkstyle/checks/design/FinalClassCheck.java#L261 paste this word. 
 
I am going to change the docs.

Reply:
>is that we cannot show violation if a superclass is inherited in a subclass that is nested otherwise it becomes non-compilable.

When making a statement like this, please always show example of failing `javac`.

Reply:
https://github.com/checkstyle/checkstyle/pull/12737#discussion_r1208538996

```
kevin@inspiron-15-5510:~/Desktop/check_style$ cat Test.java
public class Test {
    private final class SuperClass { 
        private SuperClass() {
        }

        private class SubClass extends SuperClass { 
        }
    }
}
kevin@inspiron-15-5510:~/Desktop/check_style$ javac Test.java
Test.java:6: error: cannot inherit from final Test.SuperClass
        private class SubClass extends SuperClass { 
                                       ^
1 error
kevin@inspiron-15-5510:~/Desktop/check_style$ 



```

Reply:
This resolves my question then. It seems it is an issue it doesn't examine if `superclass is inherited in a subclass`. I could have a static inner class that doesn't extend anything, or a different class.

Reply:
@rnveach Can you please give code for that you feel it is a bug.
https://github.com/checkstyle/checkstyle/pull/12737#discussion_r1205816423

Reply:
I couldn't reproduce with this PR. It seems `registerNestedSubclass` is named misleading as it is checking the class matches the super class's name.
See https://github.com/Kevin222004/checkstyle/blob/f9f132c6d9c72d3bc57ec76af358bc5ea1c101f7/src/main/java/com/puppycrawl/tools/checkstyle/checks/design/FinalClassCheck.java#L313-L325 .

Documentation should probably be updated to:
`Class is with nested sub class.` => `Class is extended by another class in the same file.`
But feel free to verify it follows the code as I don't know it as well.

Examples:
````
$ cat TestClass.java
import some.other.place.ExternalClass;

public class TestClass {
    private static class SubClass extends ExternalClass {}

    private class SuperClass1 {
        private SuperClass1() {}

      private class SubClass extends ExternalClass {}
    }

    private class SuperClass2 extends SuperClass1 {}
}

$ cat TestConfig.xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
          "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
          "https://checkstyle.org/dtds/configuration_1_3.dtd">

<module name="Checker">
    <property name="charset" value="UTF-8"/>
    <property name="severity" value="warning"/>
    <property name="haltOnException" value="false"/>

    <module name="TreeWalker">
        <module name="FinalClass"></module>
    </module>
</module>

$ java -jar checkstyle-Kevin222004-pr-FinalClass-2023-05-28-all.jar -c TestConfig.xml TestClass.java
Starting audit...
[WARN] TestClass.java:4:5: Class SubClass should be declared as final. [FinalClass]
[WARN] TestClass.java:9:7: Class SubClass should be declared as final. [FinalClass]
[WARN] TestClass.java:12:5: Class SuperClass2 should be declared as final. [FinalClass]
Audit done.
````

Reply:
The code only https://github.com/Kevin222004/checkstyle/blob/f633d9b61116664b6c00154ad96f6633e0addea2/src/main/java/com/puppycrawl/tools/checkstyle/checks/design/FinalClassCheck.java#L265-L266

> Since we are renaming lots in this class, then it seems it might be a good idea to rename method as well.

@rnveach can you suggest a good name for it
What about `registerExtendedClass`

Reply:
It sounds good to me right now.

Reply:
@rnveach I have changed the name of method as  `registerExtendedClass` and docs as well as
`Class is extended by another class in the same file.`